### ISSUE TITLE
VarChar encoding problem.

### ISSUE THREAD

**Issue by heroboy** [tediousjs/tedious#723]
see this:
https://github.com/tediousjs/tedious/blob/c22ed1866fafca127f7cf844f65b022380cfd7e9/src/data-types/varchar.js#L55-L68

You use the ascii encoding,  is it incorrect?
I see you consider the encoding in `value-parser.js`.
So I find that the input parameter encoding is not correct, the select result is correct.

**Comment by heroboy** [comment]
maybe relate to https://github.com/tediousjs/tedious/pull/113 , but it should use `iconv-lite` to encode string.

**Comment by Hadis-Fard** [comment]
@heroboy varchar stores 1 byte (ascii) characters, there's an equivalent function in nvarchar datatype object  to handle the Unicode characters
https://github.com/tediousjs/tedious/blob/85d96fee8fd3c39b08513c64deff40f4141d9e62/src/data-types/nvarchar.js#L55-L62

If you have faced an issue with encoding would you please send a repo?

**Comment by heroboy** [comment]
hi @Hadis-Fard,
Please try this:
```javascript
db.acquire((err, conn) =>
{
	if (err) return;

	var request = new tedious.Request("select '中文',@v1,@v2,@v3; set @v4 = '中文';", function (err)
	{
		if (err)
		{
			console.log(err);
			return;
		}
		conn.release();
	});
	const iconv = require('iconv-lite');
	request.on('row', cols =>
		console.log(cols.map(x => x.value))
	);
	request.on('returnValue', (name, value) =>
	{
		console.log(name, value);
	})
	console.log(iconv.encode('中文', 'cp936'));
	request.addParameter('v1', tedious.TYPES.VarChar, '中文')
	request.addParameter('v2', tedious.TYPES.NVarChar, '中文')
	request.addParameter('v3', tedious.TYPES.VarChar, iconv.encode('中文', 'cp936'),{length:128})
	request.addOutputParameter('v4', tedious.TYPES.VarChar, null);
	conn.execSql(request);
})
```
The output is 
```
[ '中文', '-', '中文', '����' ]
v4 中文
```

I tried use buffer as the parameter value,but it didn't work. Becuase in `VarChar.prototype.validate` you convert it to string if it is not string. I changed it to 
```
validate: function validate(value) {
    if (value == null) {
      return null;
    }
    if (typeof value !== 'string' && !Buffer.isBuffer(value)) { // I added this isBuffer
      if (typeof value.toString !== 'function') {
        return TypeError('Invalid string.');
      }
      value = value.toString();
    }
    return value;
  }
```
And the output would be
```
[ '中文', '-', '中文', '中文' ]
v4 中文
```
The `v3` become correct

**Comment by heroboy** [comment]
And I don't understand what you mean about `varchar stores 1 byte (ascii) characters`. As I know varchar is a byte string, but I am not sure the encoding is depend on SQLServer's settings or client OS settings or something else. Maybe I guess SQLServer's codepage setting is used to convert varchar to nvarchar, otherwise the varchar is always a byte string, what it is is explained by the client side.
Sorry my English is not good. Hope you undenstand what I was saying.

**Comment by Hadis-Fard** [comment]
as it is mentioned in [MSDN](https://docs.microsoft.com/en-us/sql/t-sql/data-types/char-and-varchar-transact-sql) varchar/char is[ non-Unicode](https://www.asciitable.com/) string data, and [nvarchar/nchar](https://docs.microsoft.com/en-us/sql/t-sql/data-types/nchar-and-nvarchar-transact-sql) is unicode string data. so when you have a _Unicode string_ data you should use _nvarchar_. 

Also we want to validate the _char_ parameters to be string or if it is possible we convert it to string
for the case of ' iconv.encode('中文', 'cp936')' you should use VarBinary type, as the iconv returns buffer.

**Comment by heroboy** [comment]
What I do `iconv.encode('中文', 'cp936')` is bacuse varchar.js doesn't correctly encode the string，I manualy do it. Not becuase it is a varbinary. So please understand Unicode, encoding, string, byte, char. You English speaking people seldom suffer from it.

**Comment by heroboy** [comment]
So please at least make the parameter value can be a buffer. It's simple to fix.

**Comment by v-suhame** [comment]
Using collation received in `ENVCHANGE` wouldn't be an ideal solution, as it's possible to have different collation for column, database and SQL Server 🤔

**Comment by v-suhame** [comment]
Both ADO and JDBC driver are sending string as NVarchar by default. Here is the TSQL sent by the drivers (captured via SQL Profiler)
```sql
-- JDBC
exec sp_executesql N'INSERT INTO c936_table (c1) VALUES(@P0)',N'@P0 nvarchar(4000)',N' 中文'

--ADO
exec sp_executesql N'INSERT INTO [dbo].[c936_table] ([c1]) VALUES (@c1);',N'@c1 nvarchar(11)',@c1=N'中文'

--tedious
exec sp_executesql @statement=N'insert into c936_table values (@c1);',@params=N'@c1 varchar(2)',@c1='-‡'
```

I was sending value '中文' as string/varchar in all 3 cases, and this was the result. Passing string as NVarchar by default seems to be a better alternative than using database collation on all of the char/varchar columns. Using NVarchar too comes at a cost, but considering that collation can be set at any level (expression, column, database or Server) its better to have a solution to handle them all.

**Comment by seandongjie** [comment]
@v-suhame 
You are right to me. Thanks.

**Comment by arthurschreiber** [comment]
I'm reopening because this is still an issue - see the discussion in https://github.com/tediousjs/tedious/issues/1294.

@heroboy You posted some code to reproduce this behaviour over in https://github.com/tediousjs/tedious/issues/723#issuecomment-378097495. Do I need to tweak any additional server or database settings (e.g. change the server collation) to reproduce this correctly? 🙇‍♂️

**Comment by heroboy** [comment]
@arthurschreiber
No. I think what ever collation sql server is, it will produce incorrect output.
But if you modify the `validate` function in my sample code, and want to make this work:
```js
request.addParameter('v3', tedious.TYPES.VarChar, iconv.encode('中文', 'cp936'),{length:128})
```
You should set sql server to match the `cp936`  encoding.
